### DISCUSSION TITLE
Convert a Reader to a Builder and modify the data.

### DISCUSSION THREAD

**Discussion by MoreTore** [capnproto/capnproto-rust#509] [Q&A]
I'm struggling to figure out how to convert a Reader into a Builder.

I have a log file filled with messages labeled "Events." This log file contains sensitive user data, and I'm trying to remove it efficiently.

```rust
    // Process each message
    while let Ok(message_reader) = read_message(&mut cursor, capnp::message::ReaderOptions::default()) {
        let event = message_reader.get_root::<crate::log_capnp::event::Reader>().unwrap();

        let mut message_builder = capnp::message::Builder::new_default();
        let mut event_builder = message_builder.init_root::<crate::log_capnp::event::Builder>();

        match event.which().unwrap() {
            crate::log_capnp::event::Params(p) => {
                if let Ok(p_reader) = p {
                    // copy p_reader to message_builder and modify a couple fields to remove sensitive data. 
                }
            },
            ... // more enum types
        }
        // Write the Builder to the output writer
        write_message(&mut writer, &message_builder).unwrap();
    }
```

**Comment by MoreTore** [comment]
I figured it out:

```rust
    // Process each message
    while let Ok(message_reader) = read_message(&mut cursor, capnp::message::ReaderOptions::default()) {
         ...
        match event.which().unwrap() {
            crate::log_capnp::event::Params(p) => {
                if let Ok(p_reader) = p {
                    let mut builder: message::Builder<message::HeapAllocator> = message::Builder::new_default();
                    builder.set_root(p_reader);
                    let mut new_p_builder = builder.get_root::<crate::log_capnp::params::Builder>().unwrap();
                    new_p_builder .set_my_field(1234);
                    let new_p_reader: log_capnp::params::Reader = new_p_builder .into_reader();
                    // you can print the new_p_reader to prove its changed
                }
            },
            ... // more enum types
        }
        // Write the Builder to the output writer
        write_message(&mut writer, &message_builder).unwrap();
    }

**Reply by MoreTore** [reply]
is there a better way to do this?

**Reply by dwrensha** [reply]
There is not a fancy way to succinctly do something like "copy everything except fields X,Y ...". What you have looks fine.

One thing to be aware of: if the fields that you are scrubbing are pointer fields (lists, structs, strings, blobs), then you should probably copy the full message one more time to fully garbage collect the hole left by the original data. Otherwise you'll have empty space in the message, and the size of that space might leak information about data that was there. (The hole is automatically zeroed out, so you don't have to worry about the actual data escaping, just its size.)

**Reply by MoreTore** [reply]
> if the fields that you are scrubbing are pointer fields (lists, structs, strings, blobs), then you should probably copy the full message one more time to fully garbage collect the hole left by the original data. Otherwise you'll have empty space in the message, and the size of that space might leak information about data that was there. (The hole is automatically zeroed out, so you don't have to worry about the actual data escaping, just its size.)


Thank you. I understand.
